packages/web/src/uiUtils.ts

```diff
@@ -141,42 +141,64 @@ export function copy(text: string) {
 
 export type Setting<T> = readonly [T, (value: T) => void, string];
 
-export function useSetting<S>(name: string | undefined, defaultValue: S, title?: string): [S, React.Dispatch<React.SetStateAction<S>>, Setting<S>] {
-  if (name)
-    defaultValue = settings.getObject(name, defaultValue);
-  const [value, setValue] = React.useState<S>(defaultValue);
-  const setValueWrapper = React.useCallback((value: React.SetStateAction<S>) => {
-    if (name)
-      settings.setObject(name, value);
-    setValue(value);
-  }, [name, setValue]);
+export function useSetting<S>(name: string, defaultValue: S, title?: string, dontPersist?: boolean): [S, (v: S) => void, Setting<S>] {
+  const value = React.useSyncExternalStore(
+      function subscribe(callback) {
+        settings.onChangeEmitter.addEventListener(name, callback);
+        return () => settings.onChangeEmitter.removeEventListener(name, callback);
+      },
+      () => {
+        return settings.getObject(name, defaultValue);
+      }
+  );
+
+  const setValueWrapper = React.useCallback((value: S) => {
+    settings.setObject(name, value, dontPersist);
+  }, [name, dontPersist]);
+
   const setting = [value, setValueWrapper, title || name || ''] as Setting<S>;
   return [value, setValueWrapper, setting];
 }
 
 export class Settings {
+  onChangeEmitter = new EventTarget();
+
+  sessionSettings: Record<string, any> = {};
+
   getString(name: string, defaultValue: string): string {
     return localStorage[name] || defaultValue;
   }
 
   setString(name: string, value: string) {
     localStorage[name] = value;
+    this.onChangeEmitter.dispatchEvent(new Event(name));
     if ((window as any).saveSettings)
       (window as any).saveSettings();
   }
 
   getObject<T>(name: string, defaultValue: T): T {
     if (!localStorage[name])
-      return defaultValue;
+      return this.sessionSettings[name] ?? defaultValue;
```

Comment:
I think we should support `undefined` and `null` as setting values, so better make `sessionSettings` a map, and use `has()` method.

Reply:
done in 8d92f44bd1024dfec4e3c728516ec015eede5f1b